src/waggle/data/vision.py

```diff
@@ -135,34 +194,87 @@ def __enter__(self):
                 raise RuntimeError(
                     f"unable to open video capture for device {self.device!r}"
                 )
+            # spin up a thread to keep up with the camera frame rate
+            if self.enable_daemon:
+                self.daemon.start()
         self.context_depth += 1
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         self.context_depth -= 1
         if self.context_depth == 0:
+            if self.enable_daemon:
+                self.daemon_need_to_stop = True
+                self.daemon.join()
             self.capture.release()
-
-    def snapshot(self):
-        ok = self.capture.grab()
-        if not ok:
-            raise RuntimeError("failed to take snapshot")
-        timestamp = get_timestamp()
-        ok, data = self.capture.retrieve()
-        if not ok:
-            raise RuntimeError("failed to retrieve the taken snapshot")
-        return ImageSample(data=data, timestamp=timestamp, format=self.format)
-
-    def stream(self):
-        while True:
+    
+    def _run(self):
+        # we sleep slighly shorter than FPS to drain the buffer efficiently
+        fps = self.capture.get(cv2.CAP_PROP_FPS)
+        sleep = 0.01
+        if fps > 0 and fps < 100:
+            sleep = 1 / (fps + 1)
+        logging.debug(f'camera FPS is {fps}. the background thread sleeps {sleep} seconds in between grab()')
+        while not self.daemon_need_to_stop:
+            self.lock.acquire()
             ok = self.capture.grab()
             if not ok:
-                break
+                pass
+                # TODO: what should we do when grab() fails?
+                # raise RuntimeError("failed to grab a frame")
+            self.timestamp = get_timestamp()
+            self.lock.release()
+            if self._ready_for_next_frame.empty():
+                self._ready_for_next_frame.put_nowait(True)
+            time.sleep(sleep)
```

Comment:
Is there a reason the sleep is needed? I would have thought the .grab is blocking?

Reply:
When I tested this with a consumer. For some reason, the consumer never had a chance to obtain `self.lock` and retrieve a frame. I suspected that the producer always had the chance to obtain the lock and do `grab()`.

I could be wrong, but without sleep in the producer, I never succeeded on retrieving a frame at the consumer side.